[PySpark] Test Spark API with actual PySpark as backend

binste opened the pull request:
Following-up on [this comment](https://github.com/duckdb/duckdb/pull/14458#issuecomment-2426124842) from @Tishj. 

Approach:
* By setting the `USE_ACTUAL_SPARK` env variable to `true`, one can now run all Spark API tests against an actual PySpark backend.
  * E.g. `USE_ACTUAL_SPARK=true python -m pytest tests/fast/spark`
  * For local development, this would require Java and Spark to be installed
* I've also set this up as part of the `Python 3.9 Linux` workflow job so it runs on every pull request. I think with this, it's also fine that not every developer will run it against Spark in production as they can use the CI for it.
  * You can see that it uses Spark in CI as the Spark tests take >40s to complete... With DuckDB, it's around 2s ;) Locally, you can also add the `-s` argument to Pytest which captures all output and which shows some PySpark output.
* Wherever you see `USE_ACTUAL_SPARK` in the tests, it means that there is a difference between DuckDB and Spark.
  * It's not that much which is very nice! I think some of the differences are ok and with this, it should be easy to find them and to make a conscious decision if they should be fixed or not.

Some thoughts on why I went with a `spark_namespace` package:
* As @Tishj, I've also tried to overwrite the Python import system to either use PySpark or DuckDB based on a Pytest command-line argument. I did not manage to make this reliable enough so i works for all cases and won't easily break in the future.
* An alternative would have been a pytest fixture which provides this namespace. It's a reliable way but it makes the tests more verbose as we can’t just import e.g. `Row` once but have to extract it every time from the namespace provided by the fixture
* Having this separate package which abstracts away the logic allowed for only minor changes to existing code and it's reliable. As long as we always import from there, it should not happen that the wrong package is used.

Main changes to tests:
* If something is read from file, before comparing it, we need to order the rows
* `assert "column" in df` does not work with PySpark and needs to be `assert "column" in df.columns`
* imports

I chose feature as target branch as it already contains some relevant changes from other PRs

binste committed:
Working version where all tests pass with DuckDB as backend and Spark can be used through USE_ACTUAL_SPARK

binste committed:
Skip test_spark_to_csv as too many differences

binste committed:
Get all tests in test_spark_with_columns_renamed to pass with Spark

binste committed:
Get all tests in test_spark_with_columns to pass with Spark

binste committed:
Get all tests in test_spark_with_column_renamed to pass with Spark

binste committed:
Get all tests in test_spark_with_column to pass with Spark

binste committed:
Get all tests in test_spark_to_parquet to pass with Spark

binste committed:
Get all tests in test_spark_session to pass with Spark

binste committed:
Get all tests in test_spark_runtime_config to pass with Spark

binste committed:
Get all tests in test_spark_readcsv to pass with Spark

binste committed:
Get all tests in test_spark_join to pass with Spark

binste committed:
Get all tests in test_spark_group_by to pass with Spark

binste committed:
Get all tests in test_spark_functions_numeric to pass with Spark

binste committed:
Get all tests in test_spark_functions_date to pass with Spark

binste committed:
Get all tests in test_spark_filter to pass with Spark

binste committed:
Get all tests in test_spark_dataframe_sort to pass with Spark

binste committed:
Get all tests in test_spark_dataframe to pass with Spark

binste committed:
Get all tests in test_spark_column to pass with Spark

binste committed:
Get all tests in test_spark_catalog to pass with Spark

binste committed:
Get all tests in test_replace_empty_value to pass with Spark

binste committed:
Remove relative imports

binste committed:
Format code

binste committed:
Install pyspark as optional dependency for tests and set up Java and Spark in CI/CD

binste commented on tools/pythonpkg/duckdb/experimental/spark/errors/__init__.py:
```diff
@@ -64,6 +64,7 @@
     "SparkUpgradeException",
     "PySparkTypeError",
     "PySparkValueError",
+    "PySparkIndexError",
```
I had to add this so that `PySparkIndexError` is importable with `from duckdb.experimental.spark.errors import *`. I assume it was not intended to exclude it from `__all__` but leaving this comment just in case there was a reason.

Tishj replied:
No reason, likely just an oversight by the contributor responsible for adding it

binste committed:
Merge branch 'feature' into test_with_spark

binste committed:
Format code

Tishj commented on tools/pythonpkg/tests/conftest.py:
```diff
@@ -224,12 +224,18 @@ def _require(extension_name, db_name=''):
-    return spark.session.builder.master(':memory:').appName('pyspark').getOrCreate()
+
+    if USE_ACTUAL_SPARK:
+        return spark.session.builder.appName('pyspark').getOrCreate()
```
I also saw this when I was testing, I don't think we need to differentiate
`master(':memory:').appName('pyspark')` is set up to be a no-op in duckdb's spark

We can use `return spark.session.builder.appName('pyspark').getOrCreate()` for both

Tishj commented on tools/pythonpkg/tests/fast/spark/test_replace_empty_value.py:
```diff
@@ -17,37 +17,46 @@ def test_replace_empty(self, spark):
-        from duckdb.experimental.spark.sql.functions import col, when
+        from spark_namespace.sql.functions import col, when
 
         df2 = df.select([when(col(c) == "", None).otherwise(col(c)).alias(c) for c in df.columns])
```
Should we not add a `sort` to this call instead of sorting the result manually afterwards?

binste replied:
Good that you bring it up. I now switched to `sort` and it shows that Spark has a null-first rule in sorting and duckdb null-last. I've added a `FIXME`.

Tishj commented on tools/pythonpkg/tests/fast/spark/test_spark_column.py:
```diff
@@ -15,11 +16,19 @@ class TestSparkColumn(object):
     def test_struct_column(self, spark):
         df = spark.createDataFrame([Row(a=1, b=2, c=3, d=4)])
```
Can you add a FIXME to this test? I think we want to fix this behavior in the future, this is exactly the type of thing that these tests against (actual) pyspark are great for finding 👍 

`FIXME: column names should be set explicitly using the Row, rather than letting duckdb assign defaults (col0, col1, etc..)`

Tishj commented on tools/pythonpkg/tests/fast/spark/test_spark_dataframe.py:
```diff
@@ -35,8 +43,11 @@ def test_dataframe_from_list_of_tuples(self, spark):
-            df = spark.createDataFrame(address, ["id", "address", "state"])
+        if not USE_ACTUAL_SPARK:
+            # Unclear why but Spark does only raise such an error, within Java,
+            # once .collect is called.
```
Thanks, I think I'm happy about throwing the error earlier, but we can add a `df.collect()` underneath so we both end up throwing (even if we throw earlier)

binste replied:
Spark throws `Py4JJavaError`, weird... I've updated the test to better show this.

Tishj commented on tools/pythonpkg/tests/fast/spark/test_spark_dataframe.py:
```diff
@@ -47,22 +58,29 @@ def test_dataframe_from_list_of_tuples(self, spark):
-        res = df.collect()
-        assert res == []
+        if not USE_ACTUAL_SPARK:
+            # Spark raises PySparkValueError [CANNOT_INFER_EMPTY_SCHEMA]
```
Can you preface this with `FIXME: `
I think we want to fix this in the future, again something we would never catch without testing against the real spark :pray:

Tishj commented on tools/pythonpkg/tests/fast/spark/test_spark_dataframe.py:
```diff
@@ -47,22 +58,29 @@ def test_dataframe_from_list_of_tuples(self, spark):
-        with pytest.raises(PySparkValueError, match="number of columns in the DataFrame don't match"):
-            df = spark.createDataFrame(address, ["id", "address"])
+        if not USE_ACTUAL_SPARK:
+            # Spark does not raise this error
```
`FIXME: `
Odd behavior actually, I guess they assign some default column name to the third one? Something I'll have to look into at some point 👍

Tishj commented on tools/pythonpkg/tests/fast/spark/test_spark_dataframe_sort.py:
```diff
@@ -59,9 +64,16 @@ def test_sort_empty_params(self, spark):
+    # See https://github.com/apache/spark/commit/0193d0f88a953063c41c41042fb58bd0badc155c
+    # for the PR which added that error to PySpark
+    @pytest.mark.skipif(
+        USE_ACTUAL_SPARK, reason="PySparkIndexError is only introduced in PySpark 4.0.0 which is not yet available"
```
We could use `hasattr` just so it switches to being tested when PySpark 4.0.0 releases
similar to this:
```py
@pytest.mark.skipif(not hasattr(pa, 'ListViewArray'), reason='The pyarrow version does not support ListViewArrays')
```

binste replied:
Good idea!

Tishj reviewed (changes requested):
Thanks for doing this work, looking at the required changes to the tests it was long overdue 😅 

I went over a couple of the changes and picked out a few minor things that stood out to me, if you can fix those up we can merge this 😄 (barring some CI difficulties with older versions, such as <https://github.com/duckdb/duckdb/actions/runs/11494990813/job/31993557125?pr=14526>)

Thanks again!

Tishj commented:
@binste 

```patch
diff --git a/tools/pythonpkg/tests/fast/spark/test_spark_functions_miscellaneous.py b/tools/pythonpkg/tests/fast/spark/test_spark_functions_miscellaneous.py
index 8802ca0ebe..18a9036407 100644
--- a/tools/pythonpkg/tests/fast/spark/test_spark_functions_miscellaneous.py
+++ b/tools/pythonpkg/tests/fast/spark/test_spark_functions_miscellaneous.py
@@ -1,8 +1,8 @@
 import pytest
 
 _ = pytest.importorskip("duckdb.experimental.spark")
-from duckdb.experimental.spark.sql import functions as F
-from duckdb.experimental.spark.sql.types import Row
+from spark_namespace.sql import functions as F
+from spark_namespace.sql.types import Row
 
 
 class TestsSparkFunctionsMiscellaneous:
diff --git a/tools/pythonpkg/tests/fast/spark/test_spark_functions_numeric.py b/tools/pythonpkg/tests/fast/spark/test_spark_functions_numeric.py
index d6d72310f1..afce98b902 100644
--- a/tools/pythonpkg/tests/fast/spark/test_spark_functions_numeric.py
+++ b/tools/pythonpkg/tests/fast/spark/test_spark_functions_numeric.py
@@ -111,7 +111,7 @@ class TestSparkFunctionsNumeric(object):
     def test_exp(self, spark):
         data = [
             (0.693,),
-            (0,),
+            (0.0,),
         ]
         df = spark.createDataFrame(data, ["firstColumn"])
         df = df.withColumn("exp_value", F.exp(F.col("firstColumn")))
@@ -148,7 +148,7 @@ class TestSparkFunctionsNumeric(object):
     def test_ln(self, spark):
         data = [
             (2.718,),
-            (1,),
+            (1.0,),
         ]
         df = spark.createDataFrame(data, ["firstColumn"])
         df = df.withColumn("ln_value", F.ln(F.col("firstColumn")))
@@ -159,7 +159,7 @@ class TestSparkFunctionsNumeric(object):
     def test_degrees(self, spark):
         data = [
             (3.14159,),
-            (0,),
+            (0.0,),
         ]
         df = spark.createDataFrame(data, ["firstColumn"])
         df = df.withColumn("degrees_value", F.degrees(F.col("firstColumn")))
```

should fix the remaining issues

binste committed:
Apply suggested code changes to fix errors

binste committed:
Simplify spark fixture

binste committed:
Apply suggestions from review (add some FIXME's etc.)

binste committed:
Format code

binste commented:
> Thanks for doing this work, looking at the required changes to the tests it was long overdue 😅
> 
> I went over a couple of the changes and picked out a few minor things that stood out to me, if you can fix those up we can merge this 😄 (barring some CI difficulties with older versions, such as https://github.com/duckdb/duckdb/actions/runs/11494990813/job/31993557125?pr=14526)
> 
> Thanks again!

Thank you for all the effort you've already put into this API! I'm really excited about it, lot of potential and a very solid foundation :) 

I've addressed all your comments. Fingers crossed that CI runs through...

Tishj reviewed (approved):
Thanks!

Mytherin commented:
Thanks!
